aws/amazon-kinesis-streams-for-fluent-bit
# Implementation of the amazon kinesis data streams fluent bit plugin

*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.


### Plugin Options

* `region`: The region which your Kinesis Data Stream is in.
* `stream`: The name of the Kinesis Data Stream that you want log records sent to. 
* `partition_key`: A partition key is used to group data by shard within a stream. A Kinesis Data Stream uses the partition key that is associated with each data record to determine which shard a given data record belongs to. For example, if your logs come from Docker containers, you can use container_id as the partition key, and the logs will be grouped and stored on different shards depending upon the id of the container they were generated from. As the data within a shard are coarsely ordered, you will get all your logs from one container in one shard roughly in order. If you don't set a partition key or put an invalid one, a random key will be generated, and the logs will be directed to random shards. If the partition key is invalid, the plugin will print an warning message.
* `data_keys`: By default, the whole log record will be sent to Kinesis. If you specify key name(s) with this option, then only those keys and values will be sent to Kinesis. For example, if you are using the Fluentd Docker log driver, you can specify data_keys log and only the log message will be sent to kinesis. If you specify multiple keys, they should be comma delimited. 
* `role_arn`: Name of an IAM role to assume (for cross account access).
* `append_newline`: If you set append_newline as true, a newline will be added after each log record.


### Test Plan

- [x] Case 1: Simple Use Case Example
    - `stream` and `region` values only
- [x] Case 2: Log Level via Env Var
    - Set log level to “debug” using FLB_LOG_LEVEL
- [x] Case 3: Data Keys Option
    - Set data_keys to “log” and check that only the log key is sent to Kinesis Data Streams.
- [x] Case 4: Assume Role
    - Use the role_arn option.
- [x] Case 5: Fluent Logger SDK with nested JSON
    - Use a Fluent Logger SDK and send a log event with at least 2 levels of nested JSON.
- [x] Case 6: SEND_FAILURE_TIMEOUT Env Var
    - Set a timeout and configure networking or permissions so that Fluent Bit can not write to Kinesis Data Streams.
- [x] Case 7: endpoint option 
- [x] Case 8: Partition Key
    - Set different values (container_id, wrong_key) for the 'partition_key' field in the OUTPUT section of the fluent-bit config file, and check to verify that logs are stored in different shards based upon the partition key value.
- [x] Case 9: Append Newline
    - Set the value for 'append_newline' filed as true in the OUTPUT section of the fluent-bit config file, and check to verify that a newline is added after each log record.
